include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+
+/**
+ * @brief Charger Interface
+ * @defgroup charger_interface Charger Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* Keep these alphabetized wrt property name */
+
+/** Runtime Dynamic Battery Parameters */
+
+/**
+ * @name Runtime Dynamic Battery Parameters
+ * @anchor CHARGER_PROPS
+ * @{
+ */
+
+/** Indicates if external supply is present for the charger */
+#define CHARGER_ONLINE	0
+/** Reports whether or not a battery is present */
+#define CHARGER_PRESENT CHARGER_ONLINE + 1
+/** Represents the charging status of the charger */
+#define CHARGER_STATUS	CHARGER_PRESENT + 1
+
+/** Reserved to demark end of common charger properties */
+#define CHARGER_COMMON_COUNT CHARGER_STATUS + 1
+/**
+ * Reserved to demark downstream custom properties - use this value as the actual value may
+ * change over future versions of this API
+ */
+#define CHARGER_CUSTOM_BEGIN CHARGER_COMMON_COUNT + 1
+
+/** Reserved to demark end of valid enum properties */
+#define CHARGER_PROP_MAX UINT16_MAX
```

Comment:
I'm mostly mimicking fuel_guage.h. I suspect this was done to be a bit more memory efficient, but I don't have any objections to using an enum as it would be easier to maintain and understand. @aaronemassey, any concerns?

Reply:
We did this because downstream users will probably want to define their own types and very well might be using Cpp in at least their client/application code. When one tries to define a custom type using as an `enum` in Cpp it fails without a cast to an `int` type.

Here's an illustrative example: 

```

enum foo {
	a = 0,
	b,
	custom = 256,
};

void do_foo(enum foo f)
{
}

int main()
{
  enum foo my_custom = custom + 1;
  // Works for C but results in a compilation failure for cpp!!!`
  //   test.cpp: In function ‘int main()’:
  //   test.cpp:15:31: error: invalid conversion from ‘int’ to ‘foo’ [-fpermissive]
  //    15 |   enum foo my_custom = custom + 1;
  //       |                        ~~~~~~~^~~
  //       |                               |
  //       |                               int
  //
}
```

Since we'd be forcing clients to do casting anyway we went with the `#define FUEL_GAUGE_PROP` pattern instead. Which is really just about as safe for C clients and won't require casting from Cpp clients.

Reply:
something like this works for me:

```diff
diff --git a/samples/cpp/cpp_synchronization/src/main.cpp b/samples/cpp/cpp_synchronization/src/main.cpp
index 3ccf39bbb2..31f9db74b4 100644
--- a/samples/cpp/cpp_synchronization/src/main.cpp
+++ b/samples/cpp/cpp_synchronization/src/main.cpp
@@ -33,6 +33,25 @@ public:
 	virtual void give(void) = 0;
 };
 
+enum charger_props {
+	CHARGER_PROP_0 = 0,
+	CHARGER_PROP_1,
+	_CHARGER_PROP_CUSTOM_START = 128,
+	_CHARGER_PROP_MAX = UINT16_MAX,
+};
+
+enum my_charger_props {
+	MY_CHARGER_PROP_0 = _CHARGER_PROP_CUSTOM_START,
+	MY_CHARGER_PROP_1,
+};
+
+typedef uint16_t charger_prop_t;
+
+static void testf(charger_prop_t prop)
+{
+	printk("prop: %u\n", prop);
+}
+
 /* specify delay between greetings (in ms); compute equivalent in ticks */
 #define SLEEPTIME  500
 #define STACKSIZE 2000
@@ -139,6 +158,11 @@ int main(void)
 			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
 	k_timer_init(&timer, NULL, NULL);
 
+	testf(CHARGER_PROP_0);
+	testf(CHARGER_PROP_1);
+	testf(MY_CHARGER_PROP_0);
+	testf(MY_CHARGER_PROP_1);
+
 	while (1) {
 		/* say "hello" */
 		printk("%s: Hello World!\n", __FUNCTION__);
```

Reply:
@gmarull @rriveramcrus 

So that works fine when invoking downstream defined `testf(charger_prop_t prop)` but if we make the same invocations using an upstream defined `testf_takes_charger_props(enum charger_props prop)` we get `error: cannot convert ‘my_charger_props’ to ‘charger_props’`.

However, I like the use of the `typedef` for `uint16_t`. **Here's an idea**: What if we keep the function declarations using `uint16_t` and switch to enums in both charger and fuel gauge? Here's an example in `main.cpp`.


```c
	testf(CHARGER_PROP_0);
	testf(CHARGER_PROP_1);
	testf(MY_CHARGER_PROP_0);
	testf(MY_CHARGER_PROP_1);

	testf_takes_charger_props(CHARGER_PROP_0);
	testf_takes_charger_props(CHARGER_PROP_1);
	// Errors
	testf_takes_charger_props(MY_CHARGER_PROP_0);
	// Errors
	testf_takes_charger_props(MY_CHARGER_PROP_1);

	// Works!
	testf_takes_uint16(CHARGER_PROP_0);
	testf_takes_uint16(CHARGER_PROP_1);
	testf_takes_uint16(MY_CHARGER_PROP_0);
	testf_takes_uint16(MY_CHARGER_PROP_1);
```

I hate the idea of using `#define PROP` since it's a bit error prone. Since this approach would work with both C/Cpp downstream clients.

Reply:
> @gmarull @rriveramcrus
> 
> So that works fine when invoking downstream defined `testf(charger_prop_t prop)` but if we make the same invocations using an upstream defined `testf_takes_charger_props(enum charger_props prop)` we get `error: cannot convert ‘my_charger_props’ to ‘charger_props’`.

well, we should never take an enum with this approach, just use typedefed `charger_prop_t` everywhere.

> 
> However, I like the use of the `typedef` for `uint16_t`. **Here's an idea**: What if we keep the function declarations using `uint16_t` and switch to enums in both charger and fuel gauge? Here's an example in `main.cpp`.

Using fixed-with integer doesn't hide what is really an implementation detail, that could be changed to anything else if needed (potentially breaking a public interface)

Reply:
Just confirming @gmarull . You're suggesting the below right?

```c
enum charger_props {
	CHARGER_PROP_0 = 0,
	CHARGER_PROP_1,
	_CHARGER_PROP_CUSTOM_START = 128,
	_CHARGER_PROP_MAX = UINT16_MAX,
};

enum downstream_charger_props {
	MY_CHARGER_PROP_0 = _CHARGER_PROP_CUSTOM_START,
	MY_CHARGER_PROP_1,
};

typedef uint16_t charger_prop_t;

static void testf(charger_prop_t prop)
{
	printk("prop: %u\n", prop);
}
```

```
int main(void)
{
	...

	testf(CHARGER_PROP_0);
	testf(CHARGER_PROP_1);
	testf(MY_CHARGER_PROP_0);
	testf(MY_CHARGER_PROP_1);
        
        ...
}
```

If so, I'm totally on board.

Reply:
Yes, I think that leads to a better usability.

Reply:
Cool.

@gmarull - I'll migrate the fuel gauge API after https://github.com/zephyrproject-rtos/zephyr/pull/57362 is merged.

@rriveramcrus - Apologies for the prior API leading you astray. I think typedefing `uint16_t` to `charger_property_t` and switching to an `enum` is the right move.